n-api: detect deadlocks in thread-safe function

gabrielschulhof opened the pull request:
We introduce status `napi_would_deadlock` to be used as a return status
by `napi_call_threadsafe_function` if the call is made with
`napi_tsfn_blocking` on the main thread and the queue is full.

Fixes: https://github.com/nodejs/node/issues/32615
Signed-off-by: @gabrielschulhof

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

addaleax commented on src/node_api.cc:
```diff
@@ -148,12 +149,15 @@ class ThreadSafeFunction : public node::AsyncResource {
 
   napi_status Push(void* data, napi_threadsafe_function_call_mode mode) {
     node::Mutex::ScopedLock lock(this->mutex);
+    uv_thread_t the_thread = uv_thread_self();
```
```suggestion
    uv_thread_t current_thread = uv_thread_self();
```

addaleax commented on doc/api/n-api.md:
```diff
@@ -5091,6 +5092,12 @@ preventing data from being successfully added to the queue. If set to
+main thread, it will return `napi_would_deadlock` if the queue is full and it
+was called with `napi_tsfn_blocking`. The reason for this is that the main
+thread is responsible for reducing the number of items in the queue so, if it
+waits for room to become available on the queue, then it will deadlock.
```
Add a `changes:` entry for `napi_create_threadsafe_function` below?

gabrielschulhof commented:
@addaleax I updated the variable name, added the `changes:` to `napi_call_threadsafe_function()` and re-worder the documentation about its return value a bit.

legendecas commented on doc/api/n-api.md:
```diff
@@ -5227,6 +5234,10 @@ This API may be called from any thread which makes use of `func`.
+changes:
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/32689
+    description: Return `napi_would_deadlock` when called from the main thread.
```
```suggestion
    description: Return `napi_would_deadlock` if called with `napi_tsfn_blocking` from the main thread and the queue is full.
```
Just another clarification...

gabrielschulhof commented:
Landed in aeb7084fe6446350ec032e9819746126811bf44f.

Himself65 commented:
just tip. downstream #32689 or #32860
